Add CLI support to generate morphometrics file.

vs74 opened the pull request:
Following discussions in issue #421, this PR aims to implement CLI to generate morphometrics file.

**CLI COMMAND :** `axondeepseg_morphometrics -i path_to_image.png  -f  axon_morphometrics.xlsx -s pixel_size`  

**DOCUMENTATION** - https://axondeepseg--434.org.readthedocs.build/en/434/documentation.html#morphometrics

Fixes #421

**Test the PR** 
Kindly refer to the [RTD](https://axondeepseg--434.org.readthedocs.build/en/434/documentation.html#morphometrics) and test the functionalities as mentioned under the **Morphometrics** section of the RTD.

**DONE: -**

- [x] Documentation: Add an example regarding usage for CLI
- [x] Feature: Should generate morphometrics from directory different from image directory
- [x] Feature: Morphometrics of images from multiple folders much like `axondeepseg` CLI.
- [x] Documentation: Morphometrics of images from multiple folders much like `axondeepseg` CLI.
- [x] Add tests

vs74 commented:
Does anyone know if there is a way to render RTD from the forked repo and include in this PR so it would be easy to review this PR ?

jcohenadad commented:
> Does anyone know if there is a way to render RTD from the forked repo and include in this PR so it would be easy to review this PR ?

link of the built RTD for the present PR is in the checks:
![image](https://user-images.githubusercontent.com/2482071/104731011-0f6c2580-5709-11eb-8ad9-c93f26ee6890.png)

vs74 commented:
@mariehbourget, I've added some literature studies here, but these are not sufficient. If you happen to know literature where axon shape being circle or ellipse are used for morphometrics, feel free to contribute in the documentation of this PR or add them in the comments, which I will add on in the RTD. Thanks!!

vs74 commented:
This PR is ready for review.

mariehbourget commented on AxonDeepSeg/morphometrics/compute_morphometrics.py:
```diff
@@ -154,33 +161,37 @@ def get_axon_morphometrics(im_axon, path_folder=None, im_myelin=None, pixel_size
-    Returns the equivalent thickness of a myelin ring around an axon of a
-    given equivalent diameter (see note [1] below). The result is in pixels.
+    Returns the  thickness of a myelin ring around an axon of a
+    given equivalent diameter (see note [1] below) or minimum axis diameter of an ellipse. The result is in pixels.
```
```suggestion
    given equivalent diameter (see note [1] below) or minor axis of an ellipse. The result is in pixels.
```

mariehbourget commented on AxonDeepSeg/morphometrics/compute_morphometrics.py:
```diff
@@ -154,33 +161,37 @@ def get_axon_morphometrics(im_axon, path_folder=None, im_myelin=None, pixel_size
     :param axonmyelin_object (skimage.measure._regionprops): object returned after
         measuring a axon with myelin labeled region
-
+   :param axon_shape: str: shape of the axon, can either be either be circle or an ellipse
```
```suggestion
   :param axon_shape: str: shape of the axon, can either be circle or ellipse
```

mariehbourget commented on AxonDeepSeg/morphometrics/compute_morphometrics.py:
```diff
@@ -265,13 +275,14 @@ def load_axon_morphometrics(path_folder):
         from axondeepseg segmentation output
     :param pred_axon: axon mask from axondeepseg segmentation output
     :param pred_myelin: myelin mask from axondeepseg segmentation output
+    :param axon_shape: str: shape of the axon, can either be either be circle or an ellipse
```
```suggestion
    :param axon_shape: str: shape of the axon, can either be circle or ellipse
```

mariehbourget commented on AxonDeepSeg/morphometrics/compute_morphometrics.py:
```diff
@@ -324,11 +335,12 @@ def save_map_of_axon_diameters(path_folder, axon_diameter_figure):
     :param pred_axon: axon mask from axondeepseg segmentation output
     :param pred_myelin: myelin mask from axondeepseg segmentation output
     :param path_folder: absolute path of folder containing pixel size file
+    :param axon_shape: str: shape of the axon, can either be either be circle or an ellipse
```
```suggestion
    :param axon_shape: str: shape of the axon, can either be circle or ellipse
```

mariehbourget commented on AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:
```diff
@@ -1,22 +1,31 @@
     morphometrics extraction of a given sample.
     :param path_img: path of the input image (microscopy sample)
     :param path_prediction: path of the segmented image (output of AxonDeepSeg)
+    :param axon_shape: str: shape of the axon, can either be either be circle or an ellipse
```
```suggestion
    :param axon_shape: str: shape of the axon, can either be circle or ellipse
```

mariehbourget commented on AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:
```diff
@@ -43,15 +52,166 @@ def launch_morphometrics_computation(path_img, path_prediction):
+
+    
+    # Setting the arguments of the saving the morphometrics in excel file
+    ap.add_argument('-s', '--sizepixel', required=False, help='Pixel size of the image(s) to segment, in micrometers. \n'+
```
```suggestion
    ap.add_argument('-s', '--sizepixel', required=False, help='Pixel size of the image(s) to compute morphometrics, in micrometers. \n'+
```

mariehbourget commented on AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:
```diff
@@ -43,15 +52,166 @@ def launch_morphometrics_computation(path_img, path_prediction):
+    ap.add_argument('-s', '--sizepixel', required=False, help='Pixel size of the image(s) to segment, in micrometers. \n'+
+                                                              'If no pixel size is specified, a pixel_size_in_micrometer.txt \n'+
+                                                              'file needs to be added to the image folder path. The pixel size \n'+
+                                                              'in that file will be used for the segmentation.',
```
```suggestion
                                                              'in that file will be used for the morphometrics computation.',
```

mariehbourget commented on AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:
```diff
@@ -43,15 +52,166 @@ def launch_morphometrics_computation(path_img, path_prediction):
+                                                              default=None)
+
+
+    ap.add_argument('-i', '--imgpath', required=True, nargs='+', help='Path to the image to be segmented ')
```
```suggestion
    ap.add_argument('-i', '--imgpath', required=True, nargs='+', help='Path to the image.')
```

mariehbourget commented on AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:
```diff
@@ -43,15 +52,166 @@ def launch_morphometrics_computation(path_img, path_prediction):
+    ap.add_argument('-i', '--imgpath', required=True, nargs='+', help='Path to the image to be segmented ')
+
+    
+    ap.add_argument('-f', '--filename', required=False,  help='Name of the excel file in which the morphometrics file will be  stored',
```
```suggestion
    ap.add_argument('-f', '--filename', required=False,  help='Name of the excel file in which the morphometrics will be stored',
```

mariehbourget commented on AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:
```diff
@@ -43,15 +52,166 @@ def launch_morphometrics_computation(path_img, path_prediction):
+            if (Path(str(current_path_target.with_suffix("")) + str(axon_suffix))).exists():
+                pred_axon = image.imread(str(current_path_target.with_suffix("")) + str(axon_suffix))
+            else: 
+                print("ERROR: Segmented axon mask is not present in the image folder  ",
```
```suggestion
                print("ERROR: Segmented axon mask is not present in the image folder. ",
```

mariehbourget commented on AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:
```diff
@@ -43,15 +52,166 @@ def launch_morphometrics_computation(path_img, path_prediction):
+                pred_axon = image.imread(str(current_path_target.with_suffix("")) + str(axon_suffix))
+            else: 
+                print("ERROR: Segmented axon mask is not present in the image folder  ",
+                                    "Please check that the axon mask is located in the image folder ",
```
```suggestion
                                    "Please check that the axon mask is located in the image folder. ",
```

mariehbourget commented on AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:
```diff
@@ -43,15 +52,166 @@ def launch_morphometrics_computation(path_img, path_prediction):
+            if (Path(str(current_path_target.with_suffix("")) + str(myelin_suffix))).exists():
+                pred_myelin = image.imread(str(current_path_target.with_suffix("")) + str(myelin_suffix))
+            else: 
+                print("ERROR: Segmented myelin mask is not present in the image folder.  ",
```
```suggestion
                print("ERROR: Segmented myelin mask is not present in the image folder. ",
```

mariehbourget commented on AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:
```diff
@@ -43,15 +52,166 @@ def launch_morphometrics_computation(path_img, path_prediction):
+                filename = filename + ".xlsx"
+            try:
+                # Export to excel
+                pd.DataFrame(x).to_excel(current_path_target.parent /  filename)
```
There is no export to csv file. A `.csv` extension in `-f` will results in an error `ValueError: No engine for filetype: 'csv' `

vs74 replied:
Good catch, I'll update the code and tests for the same.

vs74 replied:
Added test for saving filename with `.csv` extension and fixed that error.

mariehbourget commented on AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:
```diff
@@ -43,15 +52,166 @@ def launch_morphometrics_computation(path_img, path_prediction):
+                    )
+                    sys.exit(3)
+
+        
```
Empty lines to remove?

mariehbourget commented on AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:
```diff
@@ -43,15 +52,166 @@ def launch_morphometrics_computation(path_img, path_prediction):
+                        )],
+                        dtype=x.dtype)
+                    )
+
```
Empty lines to remove?

mariehbourget commented on AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:
```diff
@@ -43,15 +52,166 @@ def launch_morphometrics_computation(path_img, path_prediction):
+if __name__ == '__main__':
+    main()
+
+
```
```suggestion
```

mariehbourget commented on AxonDeepSeg/morphometrics/launch_morphometrics_computation.py:
```diff
@@ -43,15 +52,166 @@ def launch_morphometrics_computation(path_img, path_prediction):
+    main()
+
+
+
```
```suggestion
```

mariehbourget commented on docs/source/documentation.rst:
```diff
@@ -307,12 +310,124 @@ Then, use the following command::
+**Optional arguments:**
+
+-s SIZEPIXEL        Pixel size of the image(s) to segment, in micrometers. 
+                    If no pixel size is specified, a **pixel_size_in_micrometer.txt** file needs to be added to the image folder path ( that file should contain a single float number corresponding to the resolution of the image, i.e. the pixel size). The pixel size in that file will be used for the segmentation.
```
```suggestion
                    If no pixel size is specified, a **pixel_size_in_micrometer.txt** file needs to be added to the image folder path (that file should contain a single float number corresponding to the resolution of the image, i.e. the pixel size). The pixel size in that file will be used for the morphometrics computation.
```

mariehbourget commented on docs/source/documentation.rst:
```diff
@@ -307,12 +310,124 @@ Then, use the following command::
+                    **ellipse:** Axon shape is considered as an ellipse. In this case, diameter is computed using ellipse minor axis.
+                    The default axon shape is set to **circle**.
+
+-f FILENAME         Name of the excel file in which the morphometrics file will be stored.
```
```suggestion
-f FILENAME         Name of the excel file in which the morphometrics will be stored.
```

mariehbourget commented on docs/source/documentation.rst:
```diff
@@ -307,12 +310,124 @@ Then, use the following command::
+                    The default axon shape is set to **circle**.
+
+-f FILENAME         Name of the excel file in which the morphometrics file will be stored.
+                    The excel file extension can either be **.xlsx** or **.csv**.
```
The `csv` option is not working, see comment in `launch_morphometrics_computation.py`.

vs74 replied:
Fixed this.

mariehbourget commented on docs/source/documentation.rst:
```diff
@@ -307,12 +310,124 @@ Then, use the following command::
+    ...
+
+.. NOTE :: By default, AxonDeepSeg treats axon shape as **circle** and the calculation of the diameter is based on the axon area of the mask. 
+           For each axons, the equivalent diameter is computed, which is the diameter of a circle with the same area as the axon ::
```
```suggestion
           For each axons, the equivalent diameter is computed, which is the diameter of a circle with the same area as the axon. ::
```

mariehbourget commented on docs/source/documentation.rst:
```diff
@@ -307,12 +310,124 @@ Then, use the following command::
+
+
+.. NOTE :: In the literature, both equivalent diameter and ellipse minor axis are used to compute the morphometrics. 
+           Thus, depending on the usecase the user is advised to choose axon shape accordingly.
```
```suggestion
           Thus, depending on the usecase, the user is advised to choose axon shape accordingly.
```

mariehbourget commented on notebooks/04-morphometrics_extraction.ipynb:
```diff
@@ -86,8 +89,11 @@
    "metadata": {},
    "outputs": [],
    "source": [
+    "axon_shape = True #String, circle: axon shape is considered circle for computing morphometrics\n",
```
```suggestion
    "axon_shape = "circle" #String, circle: axon shape is considered circle for computing morphometrics\n",
```

mariehbourget commented on notebooks/04-morphometrics_extraction.ipynb:
```diff
@@ -86,8 +89,11 @@
    "outputs": [],
    "source": [
+    "axon_shape = True #String, circle: axon shape is considered circle for computing morphometrics\n",
+    "             #        ellipse: axon shape is considered ellipse for computing morphometrics\n",
```
```suggestion
                           # String, ellipse: axon shape is considered ellipse for computing morphometrics\n",
```

mariehbourget commented on notebooks/04-morphometrics_extraction.ipynb:
```diff
@@ -187,7 +195,9 @@
    "source": [
-    "fig = draw_axon_diameter(img,path_pred,pred_axon,pred_myelin)\n",
+
+    "fig = draw_axon_diameter(img,path_pred,pred_axon,pred_myelin, axon_shape=shape)\n",
```
```suggestion
    "fig = draw_axon_diameter(img, path_pred, pred_axon, pred_myelin, axon_shape=shape)\n",
```

mariehbourget commented on notebooks/04-morphometrics_extraction.ipynb:
```diff
@@ -207,7 +217,7 @@
    "source": [
     "# Compute aggregate metrics\n",
-    "aggregate_metrics = get_aggregate_morphometrics(pred_axon,pred_myelin,path_folder)\n",
+    "aggregate_metrics = get_aggregate_morphometrics(pred_axon,pred_myelin,path_folder, shape=shape)\n",
```
```suggestion
    "aggregate_metrics = get_aggregate_morphometrics(pred_axon, pred_myelin, path_folder, shape=shape)\n",
```

mariehbourget commented on test/morphometrics/test_compute_morphometrics.py:
```diff
@@ -228,6 +337,20 @@ def test_draw_axon_diameter_creates_file_in_expected_location(self):
+    @pytest.mark.unit
+    def test_draw_axon_diameter_creates_file_in_expected_location_with_axon_as_ellipse(self):
+        img = imageio_imread(self.test_folder_path / 'image.png')
+        path_prediction = self.test_folder_path / 'AxonDeepSeg_seg-axonmyelin.png'
```
Should we use the standard axonmyelin_suffix here for the filename?

vs74 replied:
yes, I've fixed it.

mariehbourget commented on docs/source/documentation.rst:
```diff
@@ -307,12 +310,124 @@ Then, use the following command::
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+To generate morphometrics of images which are located in different folders, specify the path of the image folders using the **-i** argumument of the CLI separated by space. For instance, to compute morphometrics of the images present in the folders **'test_sem_image/image1_sem/'** and **'test_sem_image/image2_sem/'** of the test dataset, use the following command::
+
+    axondeepseg_morphometrics -i test_sem_image/image1_sem/77.png test_sem_image/image2_sem/image.png
```
```suggestion
    axondeepseg_morphometrics -i test_segmentation/test_sem_image/image1_sem/77.png test_segmentation/test_sem_image/image2_sem/image.png
```

vs74 commented:
Thanks @mariehbourget for the review of this PR. I've addressed your review suggestions. In addition, as discussed in the last meeting I've changed the default name of the morphometrics file from `Morphometrics` to `axon_morphometrics`.

mariehbourget reviewed (commented):
I tested again with the new changes. Everything is working as expected including new csv option.
However, I could not open the notebook 04 with the following error
```
Error loading notebook
Unreadable Notebook: /home/mhbourget/axondeepseg/notebooks/04-morphometrics_extraction.ipynb NotJSONError('Notebook does not appear to be JSON: \'{\\n "cells": [\\n {\\n "cell_type": "m...')
```
I think this is not related to this PR but to #399 that should be merged first.

mathieuboudreau commented:
What’s the current statut for that PR @mariehbourget @vs74 ? Is merging #399 essential before completing this feature?

mariehbourget commented on docs/source/documentation.rst:
```diff
@@ -269,12 +272,81 @@ Then, use the following command::
+You can generate morphometrics using AxonDeepSeg via the command line interface.
+
+Syntax
+~~~~~
```
```suggestion
~~~~~~
```

mariehbourget commented on docs/source/documentation.rst:
```diff
@@ -269,12 +272,81 @@ Then, use the following command::
 
+
+Morphometrics
+---------------
```
```suggestion
-------------
```

mariehbourget reviewed (commented):
I tested the morphometrics CLI and review the documentation, everything is working as expected for 1 file of multiple files each in a different folder.

One question though, should we extend the functionality to multiple files in a single folder?
It would make sense since our CLI for segmentation has this option. But we would need a mechanism to avoid overwriting the xls file with each image.

vs74 commented:
Thanks, @mariehbourget for testing this PR. 

> One question though, should we extend the functionality to multiple files in a single folder?
> It would make sense since our CLI for segmentation has this option. But we would need a mechanism to avoid overwriting the xls file with each image.

Thanks for pointing out the possibility of the functionality of generating multiple files in a single folder I also gave a thought about it while implementing this PR, but since it required overwriting the excel files, I didn't implement it. 

Right now, I've asked the user about this [here](https://github.com/neuropoly/axondeepseg/discussions/505). Let's see what his response is. 

If the user, doesn't require generating multiple files in a single folder, then we can merge this PR and open a new issue about the additional functionality. For this, we've to fix the renaming condition of the excel files generated. For instance, an excel file could be renamed as 'image_name_axon_morphometrics.xlsx'. This would meet the user's urgent requirement. 

On the contrary, if he requires generating multiple files in a single folder, it will take 1-2 days for this PR to be ready as renaming of the morphometrics file and some tests need to added in this PR.

vs74 commented:
Looks like the user has images in different directories ( see [here](https://github.com/neuropoly/axondeepseg/discussions/505#discussioncomment-660850)). I say for now, we can merge this PR If it is alright.

I'll open a follow-up issue on adding support for generating morphometrics when all the images are within a single directory.

mariehbourget commented:
> Looks like the user has images in different directories ( see [here](https://github.com/neuropoly/axondeepseg/discussions/505#discussioncomment-660850)). I say for now, we can merge this PR if it is alright.
> 
> I'll open a follow-up issue on adding support for generating morphometrics when all the images are within a single directory.
> 
> Let me know what do you think @mariehbourget.

Yes, that is ok with me, thanks! I'll approve the PR and let you do the merge.
